app-tracking-protection/vpn-impl/src/main/java/com/duckduckgo/mobile/android/vpn/apps/ui/ExclusionListAdapter.kt

```diff
@@ -0,0 +1,318 @@
+                    setClickableLink(
+                        TrackingProtectionExclusionListActivity.LEARN_WHY_ANNOTATION,
+                        context.resources.getText(R.string.atp_ExcludedAppsDisabledLearnWhyLabel)
+                    ) { listener.onLaunchFAQ() }
+                }
+
+                BannerContent.CUSTOMISED_PROTECTION -> binding.excludedAppsEnabledVPNLabel.apply {
+                    setImageResource(com.duckduckgo.mobile.android.R.drawable.ic_info_panel_link)
+                    setText(context.resources.getString(R.string.atp_ExcludedAppsEnabledLabel))
+                }
+            }
+            binding.excludedAppsEnabledVPNLabel.minimumHeight = 0
+            binding.excludedAppsDisabledVPNLabel.apply {
+                setClickableLink(
+                    TrackingProtectionExclusionListActivity.REPORT_ISSUES_ANNOTATION,
+                    context.resources.getText(R.string.atp_ActivityDisabledLabel)
+                ) { listener.onLaunchFeedback() }
+            }
+
+            if (isListEnabled) {
+                binding.excludedAppsEnabledVPNLabel.show()
+                binding.excludedAppsDisabledVPNLabel.gone()
+            } else {
+                binding.excludedAppsEnabledVPNLabel.gone()
+                binding.excludedAppsDisabledVPNLabel.show()
+            }
+        }
+    }
+
+    private class FilterViewHolder(val binding: RowExclusionListFilterBinding) : RecyclerView.ViewHolder(binding.root) {
+        companion object {
+            fun create(parent: ViewGroup): FilterViewHolder {
+                val binding = RowExclusionListFilterBinding.inflate(LayoutInflater.from(parent.context), parent, false)
+                return FilterViewHolder(binding)
+            }
+        }
+
+        private val context: Context = binding.root.context
+
+        fun bind(
+            filterResId: Int,
+            appsFiltered: Int,
+            listener: ExclusionListListener
+        ) {
+            binding.excludedAppsFilterText.text = context.resources.getString(filterResId, appsFiltered)
+            binding.excludedAppsFilterText.setOnClickListener {
+                listener.onFilterClick(it)
+            }
+        }
+    }
+
+    private class AppViewHolder(val binding: RowExclusionListAppBinding) : RecyclerView.ViewHolder(binding.root) {
+        companion object {
+            fun create(parent: ViewGroup): AppViewHolder {
+                val binding = RowExclusionListAppBinding.inflate(LayoutInflater.from(parent.context), parent, false)
+                return AppViewHolder(binding)
+            }
+        }
+
+        private val context: Context = binding.root.context
+
+        fun bind(
+            isListEnabled: Boolean,
+            excludedAppInfo: TrackingProtectionAppInfo,
+            position: Int,
+            listener: ExclusionListListener
+        ) {
+            val appIcon = excludedAppInfo.packageIcon
```

Comment:
We need to revert this change to:
`val appIcon = itemView.context.packageManager.safeGetApplicationIcon(excludedAppInfo.packageName)`
and remove the `packageIcon` from the `TrackingProtectionAppInfo` as it invalidates the diff as it creates a new drawable object all the time. 
To check this:
Select "All", on the first item change the state of the toggle and notice the whole list is updated instead of updating only that item.